File: lib/features/auth/data/auth_repository.dart
import '../../../core/network/api_client.dart';
import '../../../core/session/session_store.dart';

class AuthRepository {
  AuthRepository({
    required ApiClient apiClient,
    required SessionStore sessionStore,
  }) : _apiClient = apiClient,
       _sessionStore = sessionStore;

  final ApiClient _apiClient;
  final SessionStore _sessionStore;

  Future<Map<String, dynamic>> login({
    required String account,
    required String password,
  }) async {
    final response = await _apiClient.post(
      '/auth/login',
      data: {
        'account': account,
        'password': password,
        'deviceName': 'flutter-app',
      },
    );
    await _persistAuthResponse(response);
    return response['user'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> register({
    String? email,
    String? phone,
    required String password,
    required String displayName,
  }) async {
    final response = await _apiClient.post(
      '/auth/register',
      data: {
        if (email != null && email.isNotEmpty) 'email': email,
        if (phone != null && phone.isNotEmpty) 'phone': phone,
        'password': password,
        'displayName': displayName,
        'deviceName': 'flutter-app',
      },
    );
    await _persistAuthResponse(response);
    return response['user'] as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> fetchCurrentUser() async {
    final response = await _apiClient.get('/users/me');
    final user = Map<String, dynamic>.from(response);
    await _sessionStore.updateUser(user);
    return user;
  }

  Future<void> logout() async {
    final refreshToken = await _sessionStore.readRefreshToken();
    if (refreshToken != null && refreshToken.isNotEmpty) {
      try {
        await _apiClient.post(
          '/auth/logout',
          data: {'refreshToken': refreshToken},
        );
      } catch (_) {
        // Keep local cleanup as the source of truth.
      }
    }

    await clearSession();
  }

  Future<void> clearSession() async {
    await _apiClient.clearSession();
  }

  Future<void> _persistAuthResponse(Map<String, dynamic> response) async {
    final accessToken = response['accessToken'] as String;
    final refreshToken = response['refreshToken'] as String;
    final user = response['user'] as Map<String, dynamic>;
    _apiClient.setAccessToken(accessToken);
    await _sessionStore.saveSession(
      accessToken: accessToken,
      refreshToken: refreshToken,
      user: user,
    );
  }
}

File: ios_native_app/ServEaseApp/Services/Repositories/AuthRepository.swift
import Foundation

actor AuthRepository {
    private let client = ApiClient.shared

    // MARK: - Login

    func login(account: String, password: String) async throws -> AuthResponse {
        let body = LoginRequest(account: account, password: password)
        let response: AuthResponse = try await client.post("auth/login", body: body)

        // Persist tokens
        try KeychainManager.shared.saveAccessToken(response.accessToken)
        try KeychainManager.shared.saveRefreshToken(response.refreshToken)
        await client.setAccessToken(response.accessToken)

        // Persist user JSON for offline
        if let data = try? JSONEncoder().encode(response.user) {
            UserDefaultsManager.shared.saveUser(data)
        }

        return response
    }

    // MARK: - Register

    func register(email: String?, phone: String?, password: String, displayName: String) async throws -> AuthResponse {
        let body = RegisterRequest(email: email, phone: phone, password: password, displayName: displayName)
        let response: AuthResponse = try await client.post("auth/register", body: body)

        try KeychainManager.shared.saveAccessToken(response.accessToken)
        try KeychainManager.shared.saveRefreshToken(response.refreshToken)
        await client.setAccessToken(response.accessToken)

        if let data = try? JSONEncoder().encode(response.user) {
            UserDefaultsManager.shared.saveUser(data)
        }

        return response
    }

    // MARK: - Current user

    func fetchCurrentUser() async throws -> User {
        let user: User = try await client.get("users/me")
        if let data = try? JSONEncoder().encode(user) {
            UserDefaultsManager.shared.saveUser(data)
        }
        return user
    }

    // MARK: - Logout

    func logout() async throws {
        if let refreshToken = try KeychainManager.shared.readRefreshToken() {
            let body = LogoutRequest(refreshToken: refreshToken)
            let _: [String: String]? = try? await client.post("auth/logout", body: body)
        }
        await clearSession()
    }

    func clearSession() async {
        await client.clearSession()
        UserDefaultsManager.shared.clearAll()
    }
}
